### PR TITLE
WasmFS: Support emscripten_wget

### PR DESCRIPTION
**Pull request by kripken**
This is one of those methods that has several intrusive API calls to the JS
API that are tricky for WasmFS. So it is simplest to just reimplement it in
wasm, as with the other methods in `system/lib/wasmfs/emscripten/`.

(I did try the other route, for comparison: https://github.com/emscripten-core/emscripten/compare/wasmfs.wget?expand=1 )

Also extend the existing test to check that we recursively create subdirs
as the docs say and the JS FS impl does.

### CODE REVIEW

File: src/library_async.js [hunk 1]
```diff
@@ -462,6 +462,7 @@ mergeInto(LibraryManager.library, {
     return Asyncify.handleSleep((wakeUp) => safeSetTimeout(wakeUp, ms));
   },
 
+#if !WASMFS
```
Comment (sbc100) [inline src/library_async.js:465]: Can we remove this function completely and always use the native version?

Having two version of this function around increases the maintenance burden.

Also, using the native version in both cases increases the amount of testing we apply to the native version.

File: system/lib/wasmfs/emscripten.cpp [hunk 2]
```diff
@@ -65,3 +67,41 @@ char *emscripten_get_preloaded_image_data_from_FILE(FILE *file,
...
+    if (result < 0 && errno != EEXIST) {
+      return;
+    }
+  }
+
+  // Fetch the data.
+  void* buffer;
+  int num;
+  int error;
+  emscripten_wget_data(url, &buffer, &num, &error);
```
Comment (sbc100) [inline system/lib/wasmfs/emscripten.cpp:93]: Why doesn't think function just return the error code I wonder?

File: system/lib/wasmfs/paths.cpp [hunk 3]
```diff
@@ -170,4 +170,34 @@ ParsedFile getFileFrom(std::shared_ptr<Directory> base, std::string_view path) {
   return doParseFile(path, base, FollowLinks, recursions);
 }
 
+std::vector<std::string_view> splitPath(std::string_view path) {
```
Comment (sbc100) [inline system/lib/wasmfs/paths.cpp:173]: Perhaps we can move this (seemingly very generic) function into `emscripten_wget.c` along with the implementation of `emscripten_wget` itself?

File: src/library_async.js (hunk 1 above)
Comment (kripken) [reply src/library_async.js]: Makes sense, in this case it looks like there's no reason not to move to C entirely now. I did not that now.

I wasn't sure which system library to put this in, HTML5 seemed closest..? Or a new one?

File: system/lib/wasmfs/emscripten.cpp (hunk 2 above)
Comment (kripken) [reply system/lib/wasmfs/emscripten.cpp:93]: Good question... no idea.

File: system/lib/wasmfs/paths.cpp (hunk 3 above)
Comment (kripken) [reply system/lib/wasmfs/paths.cpp]: (this is now moved into another system library together with `emscripten_wget`)

File: system/lib/wasmfs/emscripten.cpp [hunk 4]
```diff
@@ -65,3 +67,41 @@ char *emscripten_get_preloaded_image_data_from_FILE(FILE *file,
...
+  if (error) {
+    return;
+  }
+
+  // Write the data.
+  FILE* f = fopen(file, "w");
+  if (f) {
+    fwrite(buffer, num, 1, f);
+    fclose(f);
```
Comment (tlively) [inline system/lib/wasmfs/emscripten.cpp:102]: It might be more efficient to use `open` and `write` directly rather than stdio's buffered implementation, but then again, it might not make much difference. I guess there could be a code size difference if the user program doesn't use stdio?

File: system/lib/wasmfs/paths.cpp (hunk 3 above)
Comment (tlively) [reply system/lib/wasmfs/paths.cpp]: I'm concerned about the duplication of path parsing logic between this and `doParseParent`. Can we get away without this separate splitting logic, perhaps by looping over all the path separators in the input string in the implementation of `emscripten_wget` directly? In other words, can we implement this the same way we would if we didn't have access to WasmFS internals?

File: src/library_async.js (hunk 1 above)
Comment (sbc100) [reply src/library_async.js]: Good question.  I'd rather not add an entirely new library.. so yes html5 makes sense to me.

File: system/lib/wasmfs/paths.cpp (hunk 3 above)
Comment (sbc100) [reply system/lib/wasmfs/paths.cpp]: What we really want here is recusrive mkdir helper.  How about just copying this one.. it looks very small: 
 https://stackoverflow.com/a/2336245.. that way we can avoid using C++ at all.

File: system/lib/wasmfs/emscripten.cpp (hunk 4 above)
Comment (sbc100) [reply system/lib/wasmfs/emscripten.cpp:102]: I agree.. using lower level syscalls would probably be a good idea here.. although any program calling this function is likely to then use the higher level FS calls itself.. so maybe not worth it?

File: system/lib/wasmfs/paths.cpp (hunk 3 above)
Comment (kripken) [reply system/lib/wasmfs/paths.cpp]: @tlively This doesn't use any WasmFS internals since it was moved out in the last commits.

@sbc100 The current code is more clear than that link, I think? Maybe I just prefer C++...

File: system/lib/wasmfs/emscripten.cpp (hunk 4 above)
Comment (kripken) [reply system/lib/wasmfs/emscripten.cpp:102]: Sounds good, rewritten to use the lower-level calls. (Might not help much, but can't hurt I guess.)

**Comment by kripken** [comment]
(sorry, previously both the new code and the old code in wasmfs were still in this PR. I removed the old code now.)

File: system/lib/wasmfs/paths.cpp (hunk 3 above)
Comment (sbc100) [reply system/lib/wasmfs/paths.cpp]: I think the C version is much simpler.   It takes way fewer lines, no?  And it won't bring in libc++ code so should be a *lot* smaller.

Also, it seems reasonable to be able to use emscripten_wget in C programs without linking libc++ (know once we switch -sWASMFS by default that won't be possible any more.

Comment (kripken) [reply system/lib/wasmfs/paths.cpp]: Good point... for code size reasons we should do the simpler thing. I'll look into it.

Comment (kripken) [reply system/lib/wasmfs/paths.cpp]: Done.

File: system/lib/html5/emscripten_wget.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
+#include <emscripten.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+extern "C" {
```
Comment (sbc100) [inline system/lib/html5/emscripten_wget.cpp:8]: Can you make this file `.c` and skip these markers.

File: system/lib/html5/emscripten_wget.cpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      // Continue while we succeed in creating directories or while we see that
+      // they already exist.
+      if (result < 0 && errno != EEXIST) {
+        free(copy);
+        return;
+      }
+    }
+    c++;
+  }
+  free(copy);
```
Comment (sbc100) [inline system/lib/html5/emscripten_wget.cpp:29]: Can you split this out into `static` local function call `mkdirs` or something like that?

File: system/lib/html5/emscripten_wget.cpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  int num;
+  int error;
+  emscripten_wget_data(url, &buffer, &num, &error);
+  if (error) {
+    return;
+  }
+
+  // Write the data.
+  int fd = open(file, O_WRONLY | O_CREAT);
+  if (fd >= 0) {
```
Comment (sbc100) [inline system/lib/html5/emscripten_wget.cpp:42]: Sad that we have no way to return an error here .. .

File: tools/system_libs.py [hunk 8]
```diff
@@ -1903,7 +1903,7 @@ class libhtml5(Library):
 
   cflags = ['-Oz', '-fno-inline-functions']
   src_dir = 'system/lib/html5'
-  src_glob = '*.c'
+  src_glob = '*.c*'
```
Comment (sbc100) [inline tools/system_libs.py:1906]: Then you can revert this change too..

**Review by sbc100 — APPROVED** [review]
LGTM!  % coments

**Comment by kripken** [comment]
(feedback addressed)

File: system/lib/html5/emscripten_wget.c [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    c++;
+  }
+  free(copy);
+  return 0;
+}
+
+void emscripten_wget(const char* url, const char* file) {
+  // Create the ancestor directories.
+  if (mkdirs(file)) {
+    return;
```
Comment (sbc100) [inline system/lib/html5/emscripten_wget.c:35]: As a followup we could/should change the return value of this function to return success/failure..

Comment (kripken) [reply system/lib/html5/emscripten_wget.c:35]: That would be a breaking public API change, though?

Comment (sbc100) [reply system/lib/html5/emscripten_wget.c:35]: I don't think changing the return type from `void` to `bool` or `int` could break anything could it?

Comment (sbc100) [reply system/lib/html5/emscripten_wget.c:35]: Sorry I already did: #19747

**Comment by sbc100** [comment]
I like your version better though!: https://github.com/emscripten-core/emscripten/pull/19751
